### ISSUE TITLE
Destructuring assignment to structure/object notation asymmetry (compared to ES6)

### ISSUE THREAD

**Issue by CubOfJudahsLion** [Microsoft/TypeScript#4268]
When I perform a destructuring assignment using object notation for the left-hand side, I've noticed that I have to wrap the assignment in parenthesis to avoid a compiler error, like this:

```typescript
({a, b, c} = objectExpression);
```

(To others reading who haven't used the feature, the omission of member names causes them to default to the names of the variables -- yet another TypeScript benefit).

Reading the ES6 spec, however, I noticed that it also allows for the parenthesis surrounding just the LHS:

```javascript
({a: varForA, b: varForB, c: varForC}) = objectExpression;
```

but the TS compiler _(1.5.3)_ would not have any of it ("_error TS2364: Invalid left-hand side of assignment"_). Are there plans to cover this second form? Is the asymmetry a design choice? I ask mostly out of curiosity; I am perfectly happy with the feature as is.

Thanks for your time.

**Comment by DanielRosenwasser** [comment]
My favorite thing is to grok the ES6 spec before I go to sleep. :smile: 

So let's start off with the production in question:

*AssignmentExpression* <span class="geq">:</span> *LeftHandSideExpression* `=` *AssignmentExpression*

Now let's refer to the spec draft:

> ### <span id="sec-12.14.1" class="secnum">[12.14.1](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-assignment-operators-static-semantics-early-errors "link to this section")</span> Static Semantics: Early Errors

> *AssignmentExpression* <span class="geq">:</span> *LeftHandSideExpression* `=` *AssignmentExpression*

> -   It is a Syntax Error if *LeftHandSideExpression* is either an *ObjectLiteral* or an *ArrayLiteral* and the lexical token sequence matched by *LeftHandSideExpression* cannot be parsed with no tokens left over using *AssignmentPattern* as the goal symbol.
-   It is an early [Reference](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-reference-specification-type) Error if *LeftHandSideExpression* is neither an *ObjectLiteral* nor an *ArrayLiteral* and IsValidSimpleAssignmentTarget of *LeftHandSideExpression* is <span class="value">false</span>.

> *AssignmentExpression* <span class="geq">:</span> *LeftHandSideExpression* *AssignmentOperator* *AssignmentExpression*
-   It is an early [Reference](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-reference-specification-type) Error if IsValidSimpleAssignmentTarget of *LeftHandSideExpression* is <span class="value">false</span>.

So we need to see if the LHS is a "valid simple assignment target", and the LHS will be a parenthesized expression so we'll check the appropriate section:

> ### <span id="sec-12.2.9.3" class="secnum">[12.2.9.3](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-grouping-operator-static-semantics-isvalidsimpleassignmenttarget "link to this section")</span> Static Semantics: IsValidSimpleAssignmentTarget

> *ParenthesizedExpression* <span class="geq">:</span> `(` *Expression* `)`

> 1.  Return IsValidSimpleAssignmentTarget of *Expression*.

Ugh, ES6 spec, why don't you just tell me what I want to know? Alright, we need to go deeper.

> ### <span id="sec-12.2.0.5" class="secnum">[12.2.0.5](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-semantics-static-semantics-isvalidsimpleassignmenttarget "link to this section")</span> Static Semantics: IsValidSimpleAssignmentTarget

> *PrimaryExpression* <span class="geq">:</span>
&nbsp;&nbsp;&nbsp;&nbsp;`this`
&nbsp;&nbsp;&nbsp;&nbsp;*Literal*
&nbsp;&nbsp;&nbsp;&nbsp;*ArrayLiteral*
&nbsp;&nbsp;&nbsp;&nbsp;*ObjectLiteral*
&nbsp;&nbsp;&nbsp;&nbsp;*FunctionExpression*
&nbsp;&nbsp;&nbsp;&nbsp;*ClassExpression*
&nbsp;&nbsp;&nbsp;&nbsp;*GeneratorExpression*
&nbsp;&nbsp;&nbsp;&nbsp;*RegularExpressionLiteral*
&nbsp;&nbsp;&nbsp;&nbsp;*TemplateLiteral*

> 1.  Return **false**.

Now if you squint really hard, you can see *ObjectLiteral* in that list, which means that in ES6, this will throw a `ReferenceError`.

______________________

*TL;DR:* The existing behavior is correct and even Babel will give the error message

```
You're trying to assign to a parenthesized expression, eg. instead of `({a}) = 0` use `({a} = 0)`
```

which would probably be a nice improvement.

**Comment by danquirk** [comment]
A separate issue for a better error may be good. The current one is technically correct (the best kind of correct!) but doesn't really lead you to success:
```
a.ts(1,1): error TS2364: Invalid left-hand side of assignment expression.
a.ts(1,3): error TS2304: Cannot find name 'a'.
a.ts(1,5): error TS2304: Cannot find name 'b'.
```

**Comment by CubOfJudahsLion** [comment]
Ah. Learned more about it than I expected to, but I get it now. Thanks.

My mistake was to trust the ECMA wiki (specifically, http://wiki.ecmascript.org/doku.php?id=harmony:destructuring). Seeing the dates, it probably hasn't been updated in ages. Hah. In my defense, it was rather late. That'll teach me to post half-asleep.

Still, happy to have the feature. And TS of course.
